lib/dockerregistry/checks.go

```diff
@@ -245,3 +252,170 @@ func (check *ImageSizeCheck) Run() error {
 
 	return nil
 }
+
+// MKImageVulnCheck returns an instance of ImageVulnCheck which
+// checks against images that have known vulnerabilities.
+// nolint[funlen]
+func MKImageVulnCheck(
+	syncContext SyncContext,
+	newPullEdges map[PromotionEdge]interface{},
+	severityThreshold int,
+	customVulnProducer ImageVulnProducer,
+) *ImageVulnCheck {
+	return &ImageVulnCheck{
+		syncContext,
+		newPullEdges,
+		severityThreshold,
+		customVulnProducer,
+	}
+}
+
+// Run is a function of ImageVulnCheck and checks that none of the
+// images to be promoted have any severe vulnerabilities.
+// nolint[errcheck]
+func (check *ImageVulnCheck) Run() error {
+	var populateRequests PopulateRequests = func(
+		sc *SyncContext,
+		reqs chan<- stream.ExternalRequest,
+		wg *sync.WaitGroup) {
+		for edge := range check.PullEdges {
+			srcImage := PromotionEdge{
+				SrcRegistry: edge.SrcRegistry,
+				SrcImageTag: edge.SrcImageTag,
+			}
+			// Only check the vulnerability for the source image if it
+			// hasn't been checked already since multiple promotion
+			// edges can contain the same source image
+			if _, found := check.PullEdges[srcImage]; found {
+				continue
+			}
+			check.PullEdges[srcImage] = nil
```

Comment:
You're mutating the check.PullEdges data structure while looping over it --- instead you should use a separate map.

Reply:
yeah good call